contribs/gnostats/agent/collector.go

```diff
@@ -0,0 +1,156 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+
+	ctypes "github.com/gnolang/gno/tm2/pkg/bft/rpc/core/types"
+
+	"github.com/gnolang/gnostats/proto"
+)
+
+type collector struct {
+	// rpcClient is used by the collector to send RPC request to the Gno node
+	caller rpcClient
+}
+
+// RPCClient and RPCBatch are local interfaces that include only the
+// required methods to make them easily mockable ands allow testing
+// of the collector
+type rpcClient interface {
+	NewBatch() rpcBatch
+}
+type rpcBatch interface {
+	Status() error
+	NetInfo() error
+	NumUnconfirmedTxs() error
+	Block(*uint64) error
+	BlockResults(*uint64) error
+	Send(context.Context) ([]any, error)
+}
+
+// CollectDynamic collects dynamic info from the Gno node using RPC
+func (c *collector) CollectDynamic(ctx context.Context) (*proto.DynamicInfo, error) {
+	// Create a new batch of RPC requests
+	batch := c.caller.NewBatch()
+
+	for _, request := range [](func() error){
+		// Request Status to get address, moniker and validator info
+		batch.Status,
+		// Request NetInfo to get peers info
+		batch.NetInfo,
+		// Request NumUnconfirmedTxs to get pending txs
+		batch.NumUnconfirmedTxs,
+		// Request Block to get the last block timestamp and proposer
+		func() error { return batch.Block(nil) },
+		// Request BlockResults to get the last block number, gas used and gas wanted
+		func() error { return batch.BlockResults(nil) },
+	} {
+		if err := request(); err != nil {
+			return nil, fmt.Errorf("unable to batch request: %w", err)
+		}
+	}
+
+	// Send the batch of requests
+	results, err := batch.Send(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to send batch: %w", err)
+	}
+
+	// Cast responses to the appropriate types
+	var (
+		status  = results[0].(*ctypes.ResultStatus)
```

Comment:
Do you think we should check the cast?

Reply:
I feel like an invalid cast couldn’t happen. Personally I wouldn’t have added a check but I can add one if you want.